### ISSUE TITLE
Order of generic types for ndarray

### ISSUE THREAD

**Issue by shoyer** [numpy/numpy#16547]
Per https://github.com/python/typing/issues/516, it seems likely that we will have a choice between two ways to write type hints:
1. `ndarray[Dtype, Shape]`
2. `ndarray[Shape, Dtype]`

We'll probably have two generic arguments, but will have the freedom to choose in which order they appear. Do we have any reason to prefer one order over the other?

One weak argument in favor of `[Shape, Dtype]` is that it matches `ndarray.__new__` (but that's so rarely used that it probably doesn't matter).

**Comment by wkschwartz** [comment]
One (admittedly rather weak!) argument for putting dtype before shape is the precedent of Rust, whose [array declaration syntax](https://doc.rust-lang.org/book/ch03-02-data-types.html#the-array-type) is `[T; N]` where `T` is a type and `N` is an integer.

**Comment by eric-wieser** [comment]
@wkschwartz: C++ has the same precedence, `std::array<T, N>`

**Comment by seberg** [comment]
I have a slight tendency towards putting the dtype first as well, but no big argument to back that up.

**Comment by BvB93** [comment]
Personally I'm leaning more towards `ndarray[Dtype, Shape]` as well, 
simply because (in most cases at least) I consider the arrays dtype as the more important descriptor.

**Comment by 0az** [comment]
> But I doubt the extra sugar of using & for Intersection would really fly from a Python typing perspective.

> Thanks for the reference. There's a comment about & there, but I suppose it doesn't work well with built-in types. Of course, we're free to implement it anyway for our types if we have a convincing use case and think its a good idea.

With [PEP 604][pep-0604], there's now precedent.

Here's another proposal in the same space: why not something like this?
```py
dtype_and_shape: np.Array[np.int_ & (2, 2)]
```

[pep-0604]: https://www.python.org/dev/peps/pep-0604/

**Comment by BvB93** [comment]
> Here's another proposal in the same space: why not something like this?

The problem here is that static type checkers will be unable to understand the `x & y` syntax, 
which I'd say is a deal breaker.

I do believe that work is currently being done on an `Intersection` type, which I suspect will (eventually?) be 
followed up by an introduction of the `&` operator, just as PEP 604 did with `|` and `Union`.

**Comment by eric-wieser** [comment]
> The problem here is that static type checkers will be unable to understand the x & y syntax,
which I'd say is a deal breaker.

Do we expect static type checkers to be able to do anything with `Shape` anyway? My understanding was that in order for `Shape` to be useful, we need to either modify or write plugins for the type checkers themselves - at which point, the door is open for teaching them about `&` too.

**Comment by BvB93** [comment]
> Do we expect static type checkers to be able to do anything with `Shape` anyway?

Not yet, though this is an area which is actively being worked on (search for "tensor typing" on [typing-sig](https://mail.python.org/archives/list/typing-sig@python.org/)). The closest thing to a concrete shape type is in the [Variadic generics PEP draft](https://mail.python.org/archives/list/typing-sig@python.org/thread/SQVTQYWIOI4TIO7NNBTFFWFMSMS2TA4J/), which presents a `TypeVar`-esque object parameterized over multiple objects rather than just one.

Long story short, it would allow for a syntaxes similar to `Callable`:
* `ndarray[float64, [int]]`  # 1D
* `ndarray[float64, [int, int]]`  # 2D
* `ndarray[float64, [int, int, int]]`  # 3D

For the time being we'll have to settle for a placeholder though, `Any` probably being the wisest due to its flexibility.

**Comment by jni** [comment]
This is exciting!

For me the fact that non-typing syntax is always (?) `np.ones(shape, dtype)` suggests putting shape first. But if Eric's suggestion of `x : np.Array[float] & np.Shaped[N,M]` becomes feasible, that would be the best option imho — it's good to be able to specify them separately.

**Comment by spometun** [comment]
> Personally I'm leaning more towards `ndarray[Dtype, Shape]` as well,
> simply because (in most cases at least) I consider the arrays dtype as the more important descriptor.

I feel the same, dtype is more descriptive. There are huge amount of possible shapes while not so many dtypes

**Comment by hameerabbasi** [comment]
All the Array construction functions are usually of the signature `(shape, dtype)` such as `zeros`, `ones`, `full` (Excluding first argument), and so on. For this reason I prefer that spelling.

**Comment by BvB93** [comment]
Ok, the PR is up: https://github.com/numpy/numpy/pull/17719.

The signature is currently set to `np.ndarray[~Shape, ~DType]` after some further consideration.
The fact that this matches, for example, the `ndarray` constructor is the (marginally) strongest argument in my opinion.
